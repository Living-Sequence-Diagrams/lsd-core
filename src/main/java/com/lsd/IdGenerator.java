package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
