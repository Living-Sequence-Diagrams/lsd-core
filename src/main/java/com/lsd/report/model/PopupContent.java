package com.lsd.report.model;

import j2html.tags.ContainerTag;
import j2html.tags.UnescapedText;

import static j2html.TagCreator.*;

/**
 * Simplifies adding content that is only displayed when a hyperlink is clicked - the id used on the href must match.
 */
public class PopupContent {

    public static ContainerTag popupDiv(String id, String title, String content) {
        return div().withId(id).withClass("overlay").attr("onclick", "location.href='#!';").with(
                div().withClass("popup").attr("onclick", "event.stopPropagation();").with(
                        h2(title),
                        a().withClass("close").withHref("#!").with(new UnescapedText("&times;")),
                        div().withClass("content").with(
                                pre(content)
                        )
                )
        );
    }
}
