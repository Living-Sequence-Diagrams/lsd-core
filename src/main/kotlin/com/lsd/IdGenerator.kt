package com.lsd

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    
    private val idSupplier: Supplier<String> = if (isDeterministic)
        Supplier { counter.incrementAndGet().toString() }
    else
        Supplier { randomUUID().toString().replace("-".toRegex(), "") }


    operator fun next(): String = idSupplier.get()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}