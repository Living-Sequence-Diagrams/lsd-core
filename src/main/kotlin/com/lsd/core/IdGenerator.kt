package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}