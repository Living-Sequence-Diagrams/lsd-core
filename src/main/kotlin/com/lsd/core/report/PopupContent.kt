package com.lsd.core.report

import com.lsd.core.IdGenerator

object PopupContent {

    private val idGenerator = IdGenerator()

    /**
     * Convenience method to create the html for a hyperlink that is linked to a popup containing the provided content.
     * 
     * @param id The id is random by default and usually won't need to be overridden unless a specific id is required for some reason.
     * @param popupTitle The popupTitle is displayed as the title on the popup box containing the provided content. It's optional.
     * @param hyperlinkText The text that gets displayed on the hyperlink.
     * @param popupContent The content that will be hidden until a user clicks on the hyperlink.
     */
    @JvmStatic
    fun popupHyperlink(
        id: String = idGenerator.next(),
        popupTitle: String = "",
        hyperlinkText: String,
        popupContent: String
    ): String {
        return """
            <a href="#$id">$hyperlinkText</a>
            <div id="$id" class="overlay" onclick="location.href='#!';">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>$popupTitle</h2>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">$popupContent</div>
                </div>
            </div>
        """.trimMargin()
    }

}