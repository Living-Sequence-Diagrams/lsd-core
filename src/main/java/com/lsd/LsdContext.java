package com.lsd;

import com.lsd.diagram.ComponentDiagramGenerator;
import com.lsd.diagram.SequenceDiagramGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.parse.MessageParser;
import com.lsd.parse.Parser;
import com.lsd.parse.SynchronousResponseParser;
import com.lsd.properties.LsdProperties;
import com.lsd.report.HtmlIndexWriter;
import com.lsd.report.HtmlReportRenderer;
import com.lsd.report.HtmlReportWriter;
import com.lsd.report.model.DataHolder;
import com.lsd.report.model.Participant;
import com.lsd.report.model.Report;
import com.lsd.report.model.Scenario;

import java.nio.file.Path;
import java.util.*;

import static com.lsd.properties.LsdProperties.DETERMINISTIC_IDS;
import static com.lsd.properties.LsdProperties.MAX_EVENTS_PER_DIAGRAM;
import static java.util.stream.Collectors.toList;

public class LsdContext {

    private static final LsdContext INSTANCE = new LsdContext();

    private final List<CapturedScenario> capturedScenarios = new ArrayList<>();
    private final List<CapturedReport> capturedReports = new ArrayList<>();
    private final List<Participant> participants = new ArrayList<>();
    private final Set<String> includes = new LinkedHashSet<>();
    private final IdGenerator idGenerator = new IdGenerator(LsdProperties.getBoolean(DETERMINISTIC_IDS));
    private final List<Parser> parsers = parsers();
    private final HtmlReportWriter htmlReportWriter = new HtmlReportWriter(new HtmlReportRenderer());
    private final int maxEventsPerDiagram = LsdProperties.getInt(MAX_EVENTS_PER_DIAGRAM);

    private CapturedScenario currentScenario = new CapturedScenario();

    /**
     * Convenience method to create a singleton instance.
     * If you prefer to use other means to share a single instance or
     * want to create multiple instances then use the default constructor instead.
     *
     * @return A singleton instance of LsdContext
     */
    public static LsdContext getInstance() {
        return INSTANCE;
    }

    public void addParticipants(List<Participant> additionalParticipants) {
        participants.addAll(additionalParticipants);
    }

    public void includeFiles(Set<String> additionalIncludes) {
        includes.addAll(additionalIncludes);
    }

    public void addFact(String key, String value) {
        currentScenario.addFact(key, value);
    }

    public void capture(SequenceEvent event) {
        currentScenario.add(event);
    }

    /**
     * Allow string representations of events to be interpreted. If none of the parsers match the pattern then nothing will be captured
     *
     * @param pattern The input string that represents some event to be captured on the sequence diagram
     * @param body    The extra data associated with the event (may be null).
     */
    public void capture(String pattern, String body) {
        var event = parsers.stream()
                .map(parser -> parser.parse(pattern, body))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();

        event.ifPresent(this::capture);
    }

    public void completeScenario(String title, String description, OutcomeStatus status) {
        currentScenario.setTitle(title);
        currentScenario.setDescription(description);
        currentScenario.setOutcomeStatus(status);
        capturedScenarios.add(currentScenario);
        currentScenario = new CapturedScenario();
    }

    public Path completeReport(String title) {
        Report report = buildReport(title);
        Path path = htmlReportWriter.writeToFile(report);
        capturedReports.add(new CapturedReport(report.getTitle(), path.getFileName().toString(), report.getStatus()));
        capturedScenarios.clear();
        currentScenario = new CapturedScenario();
        return path;
    }

    public String generateReport(String title) {
        Report report = buildReport(title);
        return htmlReportWriter.writeToString(report);
    }

    public Path createIndex() {
        return HtmlIndexWriter.writeToFile(capturedReports);
    }

    public void clear() {
        idGenerator.reset();
        capturedScenarios.clear();
        capturedReports.clear();
        currentScenario = new CapturedScenario();
    }

    public IdGenerator getIdGenerator() {
        return idGenerator;
    }

    private Report buildReport(String title) {
        return Report.builder()
                .title(title)
                .status(determineOverallStatus(capturedScenarios))
                .scenarios(capturedScenarios.stream()
                        .map(capturedScenario -> Scenario.builder()
                                .title(capturedScenario.getTitle())
                                .id(idGenerator.next())
                                .status(capturedScenario.getOutcomeStatus().getCssClass())
                                .description(capturedScenario.getDescription())
                                .facts(capturedScenario.getFacts())
                                .dataHolders(capturedScenario.getSequenceEvents().stream()
                                        .filter(DataHolder.class::isInstance)
                                        .map(DataHolder.class::cast)
                                        .collect(toList()))
                                .sequenceDiagram(SequenceDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .includes(includes)
                                        .build()
                                        .diagram(maxEventsPerDiagram)
                                        .orElse(null))
                                .componentDiagram(ComponentDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .build()
                                        .diagram()
                                        .orElse(null))
                                .build())
                        .collect(toList()))
                .build();
    }

    private String determineOverallStatus(List<CapturedScenario> capturedScenarios) {
        return capturedScenarios.stream()
                .map(CapturedScenario::getOutcomeStatus)
                .sorted()
                .findFirst()
                .map(OutcomeStatus::getCssClass)
                .orElse("");
    }

    private List<Parser> parsers() {
        return List.of(
                new MessageParser(idGenerator),
                new SynchronousResponseParser(idGenerator)
        );
    }
}
