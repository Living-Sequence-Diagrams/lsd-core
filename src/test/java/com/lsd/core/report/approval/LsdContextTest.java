package com.lsd.core.report.approval;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import com.lsd.core.domain.*;
import com.lsd.core.report.PopupContent;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashSet;
import java.util.List;

import static com.lsd.core.builders.ActivateLifelineBuilder.activation;
import static com.lsd.core.builders.DeactivateLifelineBuilder.deactivation;
import static com.lsd.core.builders.MessageBuilder.messageBuilder;
import static com.lsd.core.domain.MessageType.*;
import static com.lsd.core.domain.ParticipantType.*;
import static com.lsd.core.domain.Status.*;
import static j2html.TagCreator.*;

class LsdContextTest {

    private final LsdContext lsdContext = LsdContext.getInstance();

    private final IdGenerator idGenerator = lsdContext.getIdGenerator();

    private final LinkedHashSet<String> additionalIncludes = new LinkedHashSet<>(List.of(
            "tupadr3/font-awesome-5/hamburger",
            "tupadr3/font-awesome-5/heart"
    ));

    private final Participant arnie = ACTOR.called("A", "Arnie");
    private final Participant bettie = BOUNDARY.called("Bettie");
    private final Participant cat = CONTROL.called("Cat");
    private final Participant unused = BOUNDARY.called("Unused participant");

    @BeforeEach
    public void clearContext() {
        lsdContext.clear();
        lsdContext.addParticipants(arnie, bettie, cat, unused);
    }

    @Test
    void createsReportWithScenariosAndEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Message 1").data("some data 1").type(BI_DIRECTIONAL).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("A synchronous response").from("Gamma").to("Beta").data("200 OK").type(SYNCHRONOUS_RESPONSE).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        lsdContext.capture(messageBuilder().id(nextId()).label("Sending food <$hamburger{scale=0.4}>").from(arnie).to(bettie).colour("orange").build());
        lsdContext.capture(new TimeDelay(null));
        lsdContext.capture("Sending a response from B to A [#red]", "Thank You!");
        lsdContext.capture(new NoteLeft("Friends <$heart{scale=0.4,color=red}>", null));
        lsdContext.addFact("Something to highlight", "Lorem");
        lsdContext.addFact("Something else to highlight", "amet");
        lsdContext.addFact("Something else to highlight", "consectetur");
        lsdContext.addFact("Something else to highlight", "Thank you!");
        lsdContext.completeScenario("A Warning scenario", p(
                text("A popup with a long text that needs scrolling: "),
                a().withHref("#" + "kljasdlfj").withText("click me!"),
                PopupContent.popupDiv("kljasdlfj", "I am popup", ".. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")
        ).render(), FAILURE);

        lsdContext.capture("a request from Beta to Gamma", "Please do something");
        lsdContext.capture("sync a synchronous response from Gamma to Beta", "Some Error (123456)");
        lsdContext.addFact("some important value", "123456");
        lsdContext.completeScenario("An Error scenario", "<p>Failure! Expected value to be 123 but was 123456</p>", ERROR);

        Approvals.verify(lsdContext.completeReport("Approval Report").toFile());
    }

    @Test
    void diagramIsSplitOnNewPageEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(new PageTitle("A title"));
        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A second page")));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A third page")));
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Split by NewPage Report").toFile());
    }

    @Test
    void notesCanBeAddedToParticipants() {
        lsdContext.capture(new PageTitle("Adding notes"));
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left note", null));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new TimeDelay("a few seconds later"));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new LogicalDivider("a divider"));
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new VerticalSpace(20));
        lsdContext.capture(new NoteLeft("Left of Bettie", bettie.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Bettie", bettie.getComponentName()));

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding notes to participants").toFile());
    }

    @Test
    void participantAreColoured() {
        var greg = ACTOR.called("Greg", null, "green");
        var rosie = BOUNDARY.called("Rosie", null, "red");
        var bart = CONTROL.called("Bart", null, "blue");

        lsdContext.addParticipants(greg, rosie, bart);

        lsdContext.capture(new PageTitle("Adding colours"));
        lsdContext.capture(messageBuilder().id(nextId()).from(greg).to(rosie).label("Roses are red!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(rosie).to(bart).label("Voilets are blue!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bart).to(greg).label("Grass is green?").build());

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding colours to participants").toFile());
    }

    @Test
    void lifelineCanBeActivated() {
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(activation().of(bettie).colour("red").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(deactivation().of(bettie).build());
        lsdContext.capture(activation().of(cat).colour("blue").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(deactivation().of(cat).build());

        lsdContext.completeScenario("Lifeline activation/deactivation", null, SUCCESS);

        Approvals.verify(lsdContext.completeReport("Activating lifelines").toFile());
    }

    @Test
    void generateCombinedComponentDiagramSource() {
        var dan = CONTROL.called("Dan");

        lsdContext.addParticipants(arnie, bettie, cat, dan);

        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("message1").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("message2").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message3").build());
        lsdContext.completeScenario("Scenario 1", "", SUCCESS);
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message4").build());
        lsdContext.completeScenario("Scenario 2", "", SUCCESS);
        lsdContext.completeReport("Report 1");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(arnie).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(bettie).label("message6").build());
        lsdContext.completeScenario("Scenario 3", "", SUCCESS);
        lsdContext.completeReport("Report 2");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(dan).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(dan).to(arnie).label("message6").build());
        lsdContext.completeScenario("Scenario 4", "", SUCCESS);
        lsdContext.completeReport("Report 3");

        Approvals.verify(lsdContext.completeComponentsReport("Relationships").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenThereAreNoEvents() {
        lsdContext.includeFiles(additionalIncludes);
        lsdContext.addFact("Fact", "no diagrams");
        lsdContext.completeScenario("A scenario without events", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report").toFile());
    }
    
    @Test
    void participantTypeAndAliasCanBeOverridden() {
        var fred1 = CONTROL.called("Fred");
        var fred2 = DATABASE.called("Fred", "Freddy", "purple");

        lsdContext.addParticipants(List.of(fred1, fred2, bettie));
        lsdContext.capture(messageBuilder().id(nextId()).from("Fred").to(bettie).build());
        lsdContext.completeScenario("Participant is updated", "Freddy is used", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Fred is updated").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenEventsAreCleared() {
        lsdContext.capture(messageBuilder().id(nextId()).to("A").label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).to("B").label("in").data("start another job").type(SHORT_INBOUND).build());

        lsdContext.clearScenarioEvents();
        lsdContext.completeScenario("A scenario without events (cleared down)", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report (cleared down)").toFile());
    }

    @Test
    void createsIndex() {
        lsdContext.completeScenario("Scenario 1", "Error", ERROR);
        lsdContext.completeScenario("Scenario 2", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 3", "Success", SUCCESS);
        lsdContext.completeReport("First Report (Error)");

        lsdContext.completeScenario("Scenario 1", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 2", "Success", SUCCESS);
        lsdContext.completeReport("Second Report (Warn)");

        lsdContext.completeScenario("Scenario 1", "Success", SUCCESS);
        lsdContext.completeReport("Third Report");

        lsdContext.completeReport("Fourth Report");

        Approvals.verify(lsdContext.createIndex().toFile());
    }

    @Test
    void generateReport() {
        String result = lsdContext.generateReport("Some title");

        Approvals.verify(result);
    }

    private String nextId() {
        return idGenerator.next();
    }
}