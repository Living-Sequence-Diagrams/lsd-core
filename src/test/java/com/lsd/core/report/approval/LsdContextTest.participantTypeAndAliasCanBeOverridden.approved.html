<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>Fred is updated</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/lsd-consulting/lsd-core@3/src/main/resources/static/style.min.css">
</head>
<body>
<main role="main">
    <h1 class="logo">LSD</h1>
    <h1>Fred is updated</h1>
    <article id="2">
        <span onclick="location.href='#top';" title="top of page" class="back_to_top"></span>
        <h2 class="success">Participant is updated</h2><section class="description success">
            <h3 class="success">Description</h3>
            Freddy is used
        </section>
        
        <section class="sequence diagram svg success">
            <h3 class="success">Sequence Diagram</h3>
            <figure><?xml version="1.0" encoding="us-ascii" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="205px" preserveAspectRatio="none" style="width:140px;height:205px;background:#FFFFFF;" version="1.1" viewBox="0 0 140 205" width="140px" zoomAndPan="magnify"><defs><filter height="300%" id="f53e9v6ke3q4v" width="300%" x="-1" y="-1"><feGaussianBlur result="blurOut" stdDeviation="2.0"/><feColorMatrix in="blurOut" result="blurOut2" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0"/><feOffset dx="4.0" dy="4.0" in="blurOut2" result="blurOut3"/><feBlend in="SourceGraphic" in2="blurOut3" mode="normal"/></filter></defs><g><line style="stroke:#708090;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="39" x2="39" y1="72.4883" y2="121.7988"/><line style="stroke:#708090;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="100" x2="100" y1="72.4883" y2="121.7988"/><text fill="#8392A2" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="38" x="17.5" y="69.5352">Bettie</text><path d="M19,28 L19,52 M19,40 L36,40 " fill="none" filter="url(#f53e9v6ke3q4v)" style="stroke:#7D8590;stroke-width:1.5;"/><ellipse cx="48" cy="40" fill="#FFFFFF" filter="url(#f53e9v6ke3q4v)" rx="12" ry="12" style="stroke:#7D8590;stroke-width:1.5;"/><text fill="#8392A2" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="38" x="17.5" y="134.334">Bettie</text><path d="M19,141.2871 L19,165.2871 M19,153.2871 L36,153.2871 " fill="none" filter="url(#f53e9v6ke3q4v)" style="stroke:#7D8590;stroke-width:1.5;"/><ellipse cx="48" cy="153.2871" fill="#FFFFFF" filter="url(#f53e9v6ke3q4v)" rx="12" ry="12" style="stroke:#7D8590;stroke-width:1.5;"/><text fill="#8392A2" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="47" x="74" y="69.5352">Freddy</text><path d="M82.5,20 C82.5,10 100.5,10 100.5,10 C100.5,10 118.5,10 118.5,20 L118.5,46 C118.5,56 100.5,56 100.5,56 C100.5,56 82.5,56 82.5,46 L82.5,20 " fill="#800080" filter="url(#f53e9v6ke3q4v)" style="stroke:#7D8590;stroke-width:1.5;"/><path d="M82.5,20 C82.5,30 100.5,30 100.5,30 C100.5,30 118.5,30 118.5,20 " fill="none" style="stroke:#7D8590;stroke-width:1.5;"/><text fill="#8392A2" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="47" x="74" y="134.334">Freddy</text><path d="M82.5,147.2871 C82.5,137.2871 100.5,137.2871 100.5,137.2871 C100.5,137.2871 118.5,137.2871 118.5,147.2871 L118.5,173.2871 C118.5,183.2871 100.5,183.2871 100.5,183.2871 C100.5,183.2871 82.5,183.2871 82.5,173.2871 L82.5,147.2871 " fill="#800080" filter="url(#f53e9v6ke3q4v)" style="stroke:#7D8590;stroke-width:1.5;"/><path d="M82.5,147.2871 C82.5,157.2871 100.5,157.2871 100.5,157.2871 C100.5,157.2871 118.5,157.2871 118.5,147.2871 " fill="none" style="stroke:#7D8590;stroke-width:1.5;"/><polygon fill="#7D8590" points="50.5,99.7988,40.5,103.7988,50.5,107.7988,46.5,103.7988" style="stroke:#7D8590;stroke-width:1.0;"/><line style="stroke:#7D8590;stroke-width:1.5;" x1="44.5" x2="99.5" y1="103.7988" y2="103.7988"/><a href="#1" target="_top" title="" xlink:actuate="onRequest" xlink:href="#1" xlink:show="new" xlink:title="" xlink:type="simple"><text fill="#656D76" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="16" x="56.5" y="99.0566">#1</text></a></g></svg></figure>
            <div onclick="location.href='#!';" id="3" class="overlay">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>PlantUml markup</h2>
                    <span class="copy" title="copy" onclick="copyToClipboard('sequence_uml_code3')"></span>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">
                        <pre><code id="sequence_uml_code3">@startuml

!theme lsd-light from https://raw.githubusercontent.com/lsd-consulting/lsd-core/main/src/main/resources/static

hide unlinked




boundary Bettie

database Fred as &quot;Freddy&quot; #purple



Fred -&gt; Bettie: [[#1 {} ]]


@enduml</code></pre>
                    </div>
                </div>
            </div>
            <sub><a href="#3">[source]</a></sub>
            <sub><a href="#4">[component diagram]</a></sub>
        </section>
        <!-- Popup for component diagram -->
        <div onclick="location.href='#!';" id="4" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Component Diagram</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre><?xml version="1.0" encoding="us-ascii" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="153px" preserveAspectRatio="none" style="width:96px;height:153px;" version="1.1" viewBox="0 0 96 153" width="96px" zoomAndPan="magnify"><defs/><g><!--entity Fred--><g id="elem_Fred"><path d="M11,27 C11,17 45,17 45,17 C45,17 79,17 79,27 L79,53.0146 C79,63.0146 45,63.0146 45,63.0146 C45,63.0146 11,63.0146 11,53.0146 L11,27 " fill="#800080" style="stroke:#000000;stroke-width:1.0;"/><path d="M11,27 C11,37 45,37 45,37 C45,37 79,37 79,27 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="48" x="21" y="55.0752">Freddy</text></g><!--entity Bettie--><g id="elem_Bettie"><path d="M24.5,103 L24.5,127 M24.5,115 L41.5,115 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><ellipse cx="53.5" cy="115" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="42" x="24" y="145.0752">Bettie</text></g><path d="M45,63.2702 C45,74.2987 45,87.6182 45,98.831 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="45,98.831,49,89.831,45,93.831,41,89.831,45,98.831" style="stroke:#000000;stroke-width:1.0;"/></g></svg></pre>
                </div>
                <sub><a href="#uml4">[source]</a></sub>
            </div>
        </div>
        <!-- Popup for component diagram UML via component diagram popup -->
        <div onclick="location.href='#!';" id="uml4" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>PlantUml markup</h2>
                <span class="copy" title="copy" onclick="copyToClipboard('component_uml_code4')"></span>
                <a class="close" href="#4">&times;</a><!--show previous popup on close-->
                <div class="content">
                    <pre><code id="component_uml_code4">@startuml
!pragma layout smetana
!theme plain
skinparam backgroundColor transparent

database Fred as &quot;Freddy&quot; #purple
boundary Bettie


[Fred] --&gt; Bettie


remove @unlinked
@enduml</code></pre>
                </div>
            </div>
        </div>
        <!-- Section containing data for sequence diagram popups --><div onclick="location.href='#!';" id="1" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2></h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre></pre>
                </div>
            </div>
        </div>
    </article>
</main>
</body>
<footer>
    <script src="https://d3js.org/d3.v7.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11/build/styles/dark.min.css">
    <script src="https://cdn.jsdelivr.net/gh/lsd-consulting/lsd-core@3/src/main/resources/static/custom.min.js"></script>
    <script>
        addSliderForSvgZoom();
        highlightLifelinesWhenClicked();
    </script>
    <script src="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11/build/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js"></script>
    <script>
        highlightKeywords();
        unMarkSvg();
    </script>
</footer>
</html>
